[9.0][IMP] Correct qty_delivered for phantom bom lines

StefanRijnhart opened the pull request.

pedrobaeza commented on addons/sale_mrp/migrations/9.0.1.0/post-migrate.py:
```diff
@@ -1,6 +1,55 @@
+            sale_line.product_uom_qty
+            if delivered[sale_line.id] else 0.0)
+        if sale_line.qty_delivered != qty_delivered:
+            sale_line.qty_delivered = qty_delivered
```
Should this be done through ORM?

StefanRijnhart replied:
Yes because it ensures correct values for qty_to_invoice. I see the migration of `sale` in this edition now uses SQL to do that, and I guess we could reuse those queries but for me only a relatively small number of sale lines have phantom bom products so don't see the need to do the extra effort and run the additional risks.

StefanRijnhart replied:
How does that work btw if the migration of sale sets `qty_to_invoice` based on `qty_delivered` and only afterwards `qty_delivered` is set (using SQL) in the migration of sale_stock?

StefanRijnhart replied:
Let's ask the original author :-O https://github.com/OCA/OpenUpgrade/pull/612

StefanRijnhart replied:
I think I assumed that we would trigger a recompute of qty_to_invoice in the deferred step, and it ended up in the sale module. Shall I refactor it like that? I can then refactor the above to use SQL.

pedrobaeza replied:
OK, I see that you are touching that part. Let me know when you finish.

StefanRijnhart replied:
Well, that was solved by a simple bit of copy/paste. The only hypothetical downside is if any 3rd party or custom module depends in its own migration on correct values for sale order line's qty_to_invoice or sale order's invoice_status as they won't have been computed yet. It will be impossible to ensure this does not exist, but it will be rare I think.

StefanRijnhart replied:
Hold on, got to add a check if sale is actually installed.

StefanRijnhart replied:
OK, that's it

StefanRijnhart committed:
[IMP] Correct qty_delivered for phantom bom lines

Also, defer computation of invoicing related computed fields on sale order lines

legalsylvain reviewed (approved):
Code review. No test.

pedrobaeza reviewed (approved):
Code review (check the move of the code + new SQL for phantom BoMs)
